### PR TITLE
Todo List: Add & Remove

### PR DESCRIPTION
**Pull request by Fatima-hub333**
**Attention** (“I used “webpack-dev-server” so you have to run this command in code editor “npm run server” then paste this in your browser “http://localhost:3000/”. so you will not face that webpack issues” mentioned by the previous reviewer)
Also, I follow the given template of minimalist kindly look at that template image as well.
[Minimalist example](https://www.youtube.com/watch?v=AcUd-_Yjjqg)

<h1> Implement the following things in this task; ToDO LIST; Add & Remove;</h1>
- Remove all hardcoded items from the tasks array.
- Create a new JavaScript file for the new functionality.
- Implement a function for adding a new task (add a new element to the array).
- Implement a function for deleting a task (remove an element from the array).
- Implement a function for editing task descriptions.
- By default new tasks should have the property completed set to false and the property index set to the value of the new array length (i.e. if you're adding a 5th task to the list, the index of that task should equal to 5).
- Deleting a task should update all remaining items' indexes, so they represent the current list order and are unique(i.e. if you're deleting the first task index 1 from the list, the index of the next task(2) should set to 1)..
- All changes to the To-Do List should be saved in local storage.
- 
<h1> Implement the following things in the previous task; ToDO LIST; List structure;</h1>

- Set up a new project with webpack that is based on the webpack exercise you have already completed.
- Create an index.html file and write your HTML markup here. Create an empty To Do List placeholder.
- The index.html file must be set as a template using the HTML Webpack Plugin.

-  Create an index.js file and set an array of some simple to do tasks (array of objects). Each task object should contain three keys:

> description [string].
> completed [bool].
> index: [number].
- Write a function to iterate over the tasks array and populate an HTML list item element for each task.
- On page load render the dynamically created list of tasks in the dedicated placeholder. The list should appear in order of the index values for each task.
- Create a style.css and set rules for the To Do List. CSS must be loaded by Webpack Style/CSS Loader. Your list should be a clone of the part of the minimalist project captured in the video below.

## Thanks for taking out the time to review this project 🧨
It really is appreciated :) ❤️ 
- Note: I personally love optional suggestions so please do leave some 😄

### CODE REVIEW

File: src/modules/list-item.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
+export default class Item {
+  constructor(description, completed = false, index = new Date().getTime()) {
```
Comment (sja-thedude) [inline src/modules/list-item.js:2]: - [ ] Kindly make sure that the property `index` is set to the value of the new array length (i.e. if a 5th task is added to the list, the index of that task should equal to 5).

File: src/style.css [hunk 2]
```diff
@@ -1,48 +1,165 @@
...
+.list-title {
+  width: 100%;
   height: 40px;
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+  font-family: monospace;
+  font-size: 18px;
+  font-weight: 600;
 }
```
Comment (sja-thedude) [inline src/style.css:25]: - [ ] Kindly make sure to fix the delete function. When I select items and click on clear all selected it is not deleted the first time. You can see the video below for the error.

https://user-images.githubusercontent.com/70269507/166867148-9eedad94-165b-4d38-9dcf-6025c610f2ab.mov

- [ ] Kindly also fix the error which occurs when I tried running your app locally. Below is a screenshot of the terminal.
Screenshot 2022-05-05 at 8 20 04 AM

**Review by sja-thedude — CHANGES_REQUESTED** [review]
## Required Changes ♻️

Hi @Fatima-hub333, This is @sja-thedude your code reviewer.

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

768070f3-bc9a-4aba-b8bc-25793e00ea1e_text

## Highlights
✔️ All linters passing
✔️ Good PR title and description
✔️ Great so far implementing the add and remove for the to-do list

## Required 
- [x] Kindly update your name in the Readme file.
- [ ] _Check the comments under the review._

## Optional suggestions
- [ ] You can try copying the exact same layout with spaces for the to-do list as per the design.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [@sja-thedude] in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by sja-thedude — CHANGES_REQUESTED** [review]
## Required Changes ♻️

Hi @Fatima-hub333, This is @sja-thedude your code reviewer.

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

200w

 ## Highlights
✔️ Good job so far

 ## Required Changes ♻️
- [x] Kindly make sure to implement the changes asked previously.
- [Required Change 1](https://github.com/Fatima-hub333/ToDo-List/pull/2/files#r865559187) Kindly make sure to create an array for index example:- 
`let dataList = [];
const index = (dataList) => {
  for (let i = 0; i < dataList.length; i++) { 
    dataList[i].index = i + i;
  }
  return dataList;
}; `

- [Required Change 2](https://github.com/Fatima-hub333/ToDo-List/pull/2/files#r865561400) Kindly fix the error which occurs when I tried running your app locally. Below is a screenshot of the webpage.
Screenshot 2022-05-05 at 9 01 27 AM

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [sja-thedude] in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Fatima-hub333** [comment]
> ## Required Changes ♻️
> Hi @Fatima-hub333, This is @sja-thedude your code reviewer.
> 
> Great work on making the changes requested by a previous reviewer 👏 You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.
> 
> 200w
> 
> ## Highlights
> ✔️ Good job so far
> 
> ## Required Changes ♻️
> * [x]  Kindly make sure to implement the changes asked previously.
> * [Required Change 1](https://github.com/Fatima-hub333/ToDo-List/pull/2/files#r865559187) Kindly make sure to create an array for index example:-
>   `let dataList = []; const index = (dataList) => { for (let i = 0; i < dataList.length; i++) {  dataList[i].index = i + i; } return dataList; }; `
> * [Required Change 2](https://github.com/Fatima-hub333/ToDo-List/pull/2/files#r865561400) Kindly fix the error which occurs when I tried running your app locally. Below is a screenshot of the webpage.
> 
> Screenshot 2022-05-05 at 9 01 27 AM
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me [sja-thedude] in your question so I can receive the notification.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/list-item.js (hunk 1 above)
Comment (omar-muhamad) [reply src/modules/list-item.js:2]: - [ ] As the previous reviewer mentioned please make your new task index equal to the length of the array (i.e. if a 5th task is added to the list, the index of that task should equal to 5).

image

File: src/modules/list.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  addItem(item) {
+    this.#addItem(item);
+    this.#updateStorage();
+  }
+
+  removeItem(item) {
+    this.#removeItem(item);
+    this.#updateStorage();
+  }
```
Comment (omar-muhamad) [inline src/modules/list.js:45]: - [ ] I can see you implemented the `removeItem` function but you didn't add a button or link for each task to make you able to remove every task indvidually so please fix that by adding a button and event listener for removing a single task.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @Fatima-hub333,

Nice job, you are almost there.



## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- Github flow is used correctly. ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below.

- [ ] As the previous reviewer mentioned please fix the warnings that appear when you run the dev server.

image

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: src/modules/list.js (hunk 3 above)
Comment (Fatima-hub333) [reply src/modules/list.js:45]: As I mentioned about this in my PR there is no specific requirement in this task for index list items. Kindly have a look at this image. I think it's unnecessary to ask the same thing again & again also I explained everything.

[Minimalist example](https://www.youtube.com/watch?v=AcUd-_Yjjqg)

**Comment by mke2111** [comment]
Project Under Second Opinion Review.
Kindly wait for a response. 😄

File: src/modules/list.js (hunk 3 above)
Comment (omar-muhamad) [reply src/modules/list.js:45]: Hi @Fatima-hub333 
Kindly It's a project requirement to implement remove single-task functionality and you can see the [design ](https://web.archive.org/web/20180320194056/http://www.getminimalist.com:80/) given for this project when you click on the task to edit it the delete button for single task appeared like the image below.

image

**Review by sinansevgi — COMMENTED** [review]
Hello @Fatima-hub333  👋,

_This is an additional review as a result of the second opinion request._

- [ ] For this milestone, you should implement the CRUD (create, update, delete) methods for your application. You already implemented the method but there is no way to trigger the method from the UI. If you can watch the [example project video again](https://www.youtube.com/watch?v=AcUd-_Yjjqg), you can see that the remove button appears when you are editing the task. 

- [ ] [Project requirements](https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md) specifically ask you to set the `index` value of new items to the length of the list array. Using DateTime may look easier, but when I open two browser tabs and try to submit two different items same time, it overrides each other and one of the to-do items that I want to create disappears.  (because it tries to create two items with the same id) It may not seems like a big issue since your to-do items are stored in local storage. But if you are designing an online project, this can be considered a design flaw that can cause users to lose data.

- [ ] The warning messages that previous code reviewers mentioned can be easily eliminated by setting the webpack server to development mode. You can change `"start": "webpack serve --open",` line to `    "start": "webpack serve --open --mode development",` in your package.json file to set webpack server to development mode.

Please make these changes and request another code review.

Keep rocking 🚀!

**Pull request closed without merging**